src/utils/utils_cleaner.c

```diff
@@ -203,12 +215,18 @@ static void _ocf_cleaner_complete_req(struct ocf_request *req)
 
 	OCF_DEBUG_MSG(req->cache, "All cleaning request completed");
 
-	/* Only master contains completion function and priv */
-	cmpl = master->master_io_req;
-	cmpl(master->priv, master->error);
+	if (master->complete_queue) {
+		ocf_req_get(master);
+		ocf_engine_push_req_front_cb(master,
+				_ocf_cleaner_complete, true);
+	} else {
+		/* Only master contains completion function and priv */
+		cmpl = master->master_io_req;
+		cmpl(master->priv, master->error);
 
-	/* For additional get on master allocation */
-	ocf_req_put(master);
+		/* For additional get on master allocation */
+		ocf_req_put(master);
+	}
```

Comment:
I wonder how much of an optimization would be
```
if (req != master && master->complete_queue)
    push_req_front()
else
    cmpl()
```
to avoid redundant locking